topology/probes/vpp/vpp.go

```diff
@@ -0,0 +1,410 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package vpp
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"git.fd.io/govpp.git"
+	"git.fd.io/govpp.git/api"
+	"git.fd.io/govpp.git/core"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/skydive-project/skydive/common"
+	"github.com/skydive-project/skydive/config"
+	"github.com/skydive-project/skydive/graffiti/graph"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/topology"
+	"github.com/skydive-project/skydive/topology/probes/vpp/bin_api/interfaces"
+	"github.com/skydive-project/skydive/topology/probes/vpp/bin_api/vpe"
+)
+
+const (
+	// VPPPollingTime in milliseconds
+	VPPPollingTime = 200
+)
+
+// Probe is VPP probe
+type Probe struct {
+	sync.Mutex
+	graph        *graph.Graph                              // the graph
+	shm          string                                    // connect SHM path
+	conn         *core.Connection                          // VPP conection
+	interfaceMap map[uint32]*interfaces.SwInterfaceDetails // MAP of VPP interfaces
+	root         *graph.Node                               // root node for ownership
+
+	notifChan chan api.Message // notification channel on interfaces events
+	state     int64            // state of the probe (running or stopped)
+	wg        sync.WaitGroup   // goroutines wait group
+}
+
+func interfaceMAC(mac []byte) string {
+	s := ""
+	for i, m := range mac {
+		if i != 0 {
+			s += ":"
+		}
+		s += fmt.Sprintf("%02x", m)
+	}
+	return s
+}
+
+func interfaceDuplex(duplex uint8) string {
+	switch duplex {
+	case 1:
+		return "half"
+	case 2:
+		return "full"
+	}
+	return "unknown"
+}
+
+func interfaceUpDown(updown uint8) string {
+	if updown == 1 {
+		return "UP"
+	}
+	return "DOWN"
+}
+
+func (p *Probe) getInterfaceVrfID(ch api.Channel, index uint32) int64 {
+	req := &interfaces.SwInterfaceGetTable{SwIfIndex: index}
+	msg := &interfaces.SwInterfaceGetTableReply{}
+	err := ch.SendRequest(req).ReceiveReply(msg)
+	if err != nil {
+		logging.GetLogger().Error(err)
+		return -1
+	}
+	return int64(msg.VrfID)
+}
+
+func (p *Probe) getInterface(index uint32) *graph.Node {
+	p.graph.RLock()
+	defer p.graph.RUnlock()
+	return p.graph.LookupFirstNode(graph.Metadata{"IfIndex": int64(index), "Type": "vpp"})
+}
+
+func (p *Probe) createOrUpdateInterface(ch api.Channel, intf *interfaces.SwInterfaceDetails) *graph.Node {
+	metadata := graph.Metadata{"IfIndex": int64(intf.SwIfIndex), "Type": "vpp"}
+
+	g := p.graph
+	g.Lock()
+	defer g.Unlock()
+
+	var err error
+	node := p.graph.LookupFirstNode(metadata)
+	if node == nil {
+		node, err = g.NewNode(graph.GenID(), metadata)
+		if err != nil {
+			logging.GetLogger().Error(err)
+			return nil
+		}
+		if _, err = g.Link(p.root, node, graph.Metadata{"RelationType": "ownership"}); err != nil {
```

Comment:
use this helper : https://github.com/skydive-project/skydive/blob/master/topology/topology.go#L46

Reply:
done